### PR TITLE
oc new-app --build-env doesn't work on templates

### LINKED ISSUE #14906: oc new-app --build-env doesn't work on templates
`oc new-app --build-env` works in conjunction with an argument such as `https://github.com/openshift/nodejs-ex`, but it doesn't work in conjunction with `-f path/to/template`.  `oc new-app --env` works as expected.

```
$ oc new-app -f https://raw.githubusercontent.com/openshift/rails-ex/master/openshift/templates/rails-postgresql.json --build-env FOO=bar -o yaml | grep -C 2 FOO
$ oc new-app -f https://raw.githubusercontent.com/openshift/rails-ex/master/openshift/templates/rails-postgresql.json --env FOO=bar -o yaml | grep -C 2 FOO
        containers:
        - env:
          - name: FOO
            value: bar
          - name: DATABASE_SERVICE_NAME
--
        containers:
        - env:
          - name: FOO
            value: bar
          - name: POSTGRESQL_USER
```

### PR DESCRIPTION
**Pull request by coreydaley**
Fixes #14906

### CODE REVIEW

**Comment by coreydaley** [comment]
[test][testextended][extended:builds(core)]

**Comment by coreydaley** [comment]
@openshift/devex ptal
@jim-minter fyi

File: pkg/generate/app/cmd/newapp.go [hunk 1]
```diff
@@ -485,6 +485,11 @@ func (c *AppConfig) buildTemplates(components app.ComponentReferences, parameter
 			// if environment variables were passed in, let's apply the environment variables
 			// to every pod template object
 			for i := range result.Objects {
+				if reflect.TypeOf(result.Objects[i]) == reflect.TypeOf(&buildapi.BuildConfig{}) {
+					buildEnv := buildutil.GetBuildConfigEnv(result.Objects[i].(*buildapi.BuildConfig))
+					buildEnv = app.JoinEnvironment(buildEnv, buildEnvironment.List())
+					buildutil.SetBuildConfigEnv(result.Objects[i].(*buildapi.BuildConfig), buildEnv)
+				}
```
Comment (bparees) [inline pkg/generate/app/cmd/newapp.go]: use a type switch like GetPodSpec does.

Comment (coreydaley) [inline pkg/generate/app/cmd/newapp.go]: updated

**Comment by coreydaley** [comment]
test/flake #14043

**Comment by coreydaley** [comment]
@bparees test case added

**Comment by bparees** [comment]
[merge]

**Comment by bparees** [comment]
[merge]

Ben Parees | OpenShift

**Comment by bparees** [comment]
[test]

**Comment by bparees** [comment]
[merge]

**Comment by bparees** [comment]
Gce Provision failed. [merge]

Ben Parees | OpenShift

**Pull request merged**